### PR TITLE
feat: tutorial notebook on sem seg.

### LINKED ISSUE #259: Add notebook showing TFSegFormer fine-tuning
Notebook already [exists](https://github.com/deep-diver/segformer-tf-transformers/blob/main/notebooks/TFSegFormer_Finetune.ipynb). Just need to contribute here. 

Cc: @deep-diver

### PR DESCRIPTION
**Pull request by sayakpaul**
Adds a tutorial notebook for semantic segmentation in TensorFlow. Inspired by [this notebook](https://github.com/deep-diver/segformer-tf-transformers/blob/main/notebooks/TFSegFormer_Finetune.ipynb). 

## TODO

- [x] Add training results
- [x] Evaluation results
- [x] Inference 
- [ ] Pipeline API (not supported for TF for semantic segmentation)

Colab Notebook: https://colab.research.google.com/drive/1GNddZjsEHlZEk6E2cLF-konqj2UW0wTi?usp=sharing

Closes #259

### CODE REVIEW

**Review by amyeroberts — APPROVED** [review]
LGTM, Thanks for adding! 

Left some small nits but otherwise looks good to go ⭐

**Comment by sayakpaul** [comment]
<p>Good point. Added the following:</p><p> </p><blockquote>Alternatively, we can also leverage&nbsp;<code>metric</code>&nbsp;to compute different quantities on the validation set in a batchwise manner:&nbsp;<code>mean_iou</code>,&nbsp;<code>mean_accuracy</code>, etc.</blockquote>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/260/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sayakpaul** [comment]
@amyeroberts thanks for your comments. Addressed them.

**Comment by sayakpaul** [comment]
@nateraw friendly ping for the review.

**Review by amyeroberts — APPROVED** [review]

**Comment by nateraw** [comment]
@sayakpaul reviewing now! Quick note - in the future, can you pls link directly to the GitHub Colab notebook instead of the one on your google drive so we can make sure we're always using latest version? In this case, the right link is [here](https://colab.research.google.com/github/sayakpaul/notebooks/blob/add/segmentation/examples/semantic_segmentation-tf.ipynb).

**Comment by sayakpaul** [comment]
> @sayakpaul reviewing now! Quick note - in the future, can you pls link directly to the GitHub Colab notebook instead of the one on your google drive so we can make sure we're always using latest version? In this case, the right link is [here](https://colab.research.google.com/github/sayakpaul/notebooks/blob/add/segmentation/examples/semantic_segmentation-tf.ipynb).

Will do.

**Comment by nateraw** [comment]
Here's the resulting [model repo ](https://huggingface.co/nateraw/mit-b0-finetuned-sidewalks)that got pushed (I did stop training early cuz I was impatient). Looks a little wonky to me

**Comment by sayakpaul** [comment]
> Here's the resulting [model repo ](https://huggingface.co/nateraw/mit-b0-finetuned-sidewalks)that got pushed (I did stop training early cuz I was impatient). Looks a little wonky to me

Probably it's because of the `PushToHub` callback?

**Comment by nateraw** [comment]
Yea I think so. Don't stress on that just wanted to point it out.

**Comment by osanseviero** [comment]
<p>I think 1 could be nice to have things where they are used</p>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/260/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Review by osanseviero — APPROVED** [review]
Very cool! 🔥 

I agree with Nate's suggestions. I also added some comments :) Do we plan to do a PyTorch equivalent notebook afterwards?

**Comment by osanseviero** [comment]
Btw, feel free to open an issue in transformers about the weird model card

**Review by nateraw — APPROVED** [review]
Once any minor changes are made, this is good to go. Thanks!!

**Comment by sayakpaul** [comment]
> Very cool! 🔥
> 
> I agree with Nate's suggestions. I also added some comments :) Do we plan to do a PyTorch equivalent notebook afterwards?

There's actually associated with this blog post: https://huggingface.co/blog/fine-tune-segformer.

**Comment by osanseviero** [comment]
> There's actually one associated with this blog post: https://huggingface.co/blog/fine-tune-segformer.

Maybe worth having it in the notebooks repo as well? Also that one is both about labeling and training, so one fully focused in how to train on sem seg. as the notebook in this PR would be nice!

**Comment by sayakpaul** [comment]
<p>I think since we're being explicit about the messaging (<em>if</em> they want to share the models and other things then only the login is needed), we can leave it as is for now. I think having the login above is useful in cases where there could be intermediate pushes to the Hub (maybe a better preprocessed dataset or something?). </p><p> </p><p>As a personal preference too, I like putting all the setup related things (installation, authentication, etc.) at the beginning of the notebook so that I can get straight to coding 👨‍💻. But this is more of a philosophical thing. </p>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/260/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sayakpaul** [comment]
<p>There's usually a warning always after you authenticate. So. </p>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/260/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sayakpaul** [comment]
<p>Just added two lines. Check. </p>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/260/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sayakpaul** [comment]
<p>Added. Check once. </p>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/260/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sayakpaul** [comment]
<p>Good point! Added a note. Check once. </p>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/260/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sayakpaul** [comment]
> > There's actually one associated with this blog post: https://huggingface.co/blog/fine-tune-segformer.
> 
> Maybe worth having it in the notebooks repo as well? Also that one is both about labeling and training, so one fully focused in how to train on sem seg. as the notebook in this PR would be nice!

Agree. Will do. Issue created: https://github.com/huggingface/notebooks/issues/267

**Comment by sayakpaul** [comment]
<p>The first epoch is definitely slower than expected. </p><p> </p><blockquote>Is this callback being called on each epoch?</blockquote><p> </p><p>The default save strategy is "epoch". But can be done w.r.t steps too as per the <a href="https://github.com/huggingface/transformers/blob/v4.24.0/src/transformers/keras_callbacks.py#L267" target="_blank">source code</a>. </p>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/260/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sayakpaul** [comment]
<p>I need to investigate it. Will keep the thread open till then. </p>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/260/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sayakpaul** [comment]
<p>Done. </p>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/260/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sayakpaul** [comment]
> Here's the resulting [model repo ](https://huggingface.co/nateraw/mit-b0-finetuned-sidewalks)that got pushed (I did stop training early cuz I was impatient). Looks a little wonky to me

@nateraw found the bug! In the `compute_metrics()` function, I was not preprocessing `per_category_accuracy` and `per_category_iou` appropriately. Fixing that solved the problem. The model card now looks good to me: https://huggingface.co/sayakpaul/mit-b0-finetuned-sidewalks-dummy. It's dummy model just trained for 2 epochs. Thanks to the [original SegFormer fine-tuning notebook](https://colab.research.google.com/drive/1MdkavsjGHYcuGyjmsf9wmeAK3WvtYLty?usp=sharing) (that uses PT and Trainer). 

Training the model for 50 epochs takes quite a bit of time. I suggest doing the following: 

* Remove the outputs from the cell housing `model.fit()`. 
* Update the `compute_metrics()` code with the fixes. 

Let me know your thoughts and I will prepare the updates and I guess we'll be good to merge. 

Cc: @osanseviero

**Comment by sayakpaul** [comment]
@nateraw a friendly ping for https://github.com/huggingface/notebooks/pull/260#issuecomment-1314744963

**Review by nateraw — APPROVED** [review]
Looks good!

**Comment by sayakpaul** [comment]
Went with the following:

* Remove the outputs from the cell housing model.fit().
* Update the compute_metrics() code with the fixes.

**Comment by nateraw** [comment]
Oh sorry, I didn't realize you were asking a question 😅  - I agree with what you went with :)
